populateDB/migrations/006_status_update_open_date.sql

```diff
@@ -0,0 +1,135 @@
+
+  if crossing_id is null then
+    raise exception 'Crossing is required';
+  end if;
+
+  select * from floods.crossing where id = crossing_id into crossing_to_update;
+
+  -- If we aren't a super admin
+  if current_setting('jwt.claims.role') != 'floods_super_admin' then
+    -- and we're trying to update the status of a crossing in a different community
+    if (array_position(crossing_to_update.community_ids, current_setting('jwt.claims.community_id')::integer) is null) then
+      raise exception 'Users can only update the status of crossings within their communities';
+    end if;
+  end if;
+
+  -- If the status reason is not null
+  if status_reason_id is not null then
+    -- but the association says it should be disabled
+    if (select rule from floods.status_association where floods.status_association.status_id = new_status_update.status_id and detail = 'reason') = 'disabled' then
+      -- we shouldn't be here, throw
+      raise exception 'Status reasons are disabled for status:  %', (select name from floods.status where id = status_id);
+    end if;
+
+    -- but the status reason is for a different status
+    if (select floods.status_reason.status_id from floods.status_reason where id = status_reason_id) != new_status_update.status_id then
+      -- we shouldn't be here, throw
+      raise exception 'This status reason is not for status:  %', (select name from floods.status where id = new_status_update.status_id);
+    end if;
+  end if;
+
+  -- If the status reason is null
+  if status_reason_id is null then
+    -- but the association says it is required
+    if (select rule from floods.status_association where floods.status_association.status_id = new_status_update.status_id and detail = 'reason') = 'required' then
+      -- we shouldn't be here, throw
+      raise exception 'Status reasons are required for status:  %', (select name from floods.status where id = status_id);
+    end if;
+  end if;
+
+  -- If the status duration is not null
+  if (open_date is not null or indefinite_closure is true) then
+    -- but the association says it should be disabled
+    if (select rule from floods.status_association where floods.status_association.status_id = new_status_update.status_id and detail = 'duration') = 'disabled' then
+      -- we shouldn't be here, throw
+      raise exception 'Open dates and indefinite closure flags are disabled for status:  %', (select name from floods.status where id = status_id);
+    end if;
+  end if;
+
+  -- If the status reason is null
+  if (open_date is null and indefinite_closure is false) then
+    -- but the association says it is required
+    if (select rule from floods.status_association where floods.status_association.status_id = new_status_update.status_id and detail = 'duration') = 'required' then
+      -- we shouldn't be here, throw
+      raise exception 'An open date or indefinite closure flag is required for status:  %', (select name from floods.status where id = status_id);
```

Comment:
Tiny nitpick, I think the term "reopen date" makes more sense for these error messages.

Reply:
I can dig it